### ISSUE TITLE
Tpetra: Reduce Compilation of ordinal-as-scalar objects

### ISSUE THREAD

**Issue by csiefer2** [trilinos/Trilinos#13094]
In order to provide an analogue for Epetra_IntVector, Tpetra rebuilds sizable portions of its library with Scalar=GO and Scalar=int (if int is not the GO).  We clearly need Vector<int> for moving around things like MPI rank ids, but it isn't obvious we need as much aGO/int-as-Scalar as we have in the code.

We'd like to (optionally) remove a bunch of this extra compilation of GO/int-as-Scalar to reduce build time / library size (as requested by @ibaned), for applications that don't need it (like his).

This can be done by mucking with the ETI logic in packages/tpetra/core/src.

One note is that Zoltan2's Hypergraph algorithm _does_ explicitly using CrsMatrix<int>, so that would have to be disabled if we go down this path. @egboman comments?

This issue does not occur in the same way for Ifpack2 or MueLu since they process ETI differently and don't do GO/int-as-scalar.

**Comment by egboman** [comment]
@csiefer2 Where do you see Zoltan2 use Tpetra CrsMatrix with integer types? I thought Zoltan2 just calls old Zoltan to use PHG, and old Zoltan does not use Tpetra or Epetra.

**Comment by cgcgcg** [comment]
@egboman Maybe this is what @csiefer2 had in mind?
https://github.com/trilinos/Trilinos/blob/ee6599d413f7ef5786a5ad1f719bbfe3a4d22300/packages/zoltan2/core/CMakeLists.txt#L38-L50

**Comment by egboman** [comment]
Maybe, but what does this have to do with hypergraphs?

**Comment by cgcgcg** [comment]
https://github.com/trilinos/Trilinos/blob/ee6599d413f7ef5786a5ad1f719bbfe3a4d22300/packages/zoltan2/core/src/models/Zoltan2_HyperGraphModel.hpp#L540
https://github.com/trilinos/Trilinos/blob/ee6599d413f7ef5786a5ad1f719bbfe3a4d22300/packages/zoltan2/core/src/models/Zoltan2_ModelHelpers.hpp#L60-L64

**Comment by egboman** [comment]
@csiefer2 Maybe. This needs to be done carefully. What if somebody actually wants to use Zoltan2 with hypergraph partitioning but has no clue about Tpetra and scalar types? How would they know to build Trilinos with SC=int? Would they get an error at compile time or run time?

**Comment by egboman** [comment]
@csiefer2 

I think (1) is the safest option, but might not help you much reduce compilation time. Option (2) also seems OK but adds another option.

**Comment by egboman** [comment]
Yes, looks good to me! Note a typo "wwhen" in Zoltan_config.h :)
